### PR TITLE
support config QoSClass/PriorityClass/ColocationPriority in Profile

### PR DESCRIPTION
**Pull request by eahydra**
Signed-off-by: Tao Li <joseph.t.lee@outlook.com>

### Ⅰ. Describe what this PR does

Support config QoSClass/PriorityClass/ColocationPriority in Profile.
And webhook mutating Pod inject QoSClass etc, validating Pod without QosClass and so on.

### CODE REVIEW

File: apis/config/v1alpha1/cluster_colocation_profile_types.go [hunk 1]
```diff
@@ -52,6 +50,25 @@ type ClusterColocationProfileSpec struct {
 	// +optional
 	SchedulerName string `json:"schedulerName,omitempty"`
 
+	// QoSClass describes the type of koordinator QoS that the Pod is running.
+	// Can be LSE/LSR/LS/BE/SYSTEM.
```
Comment (jasonliu747) [inline apis/config/v1alpha1/cluster_colocation_profile_types.go:54]: consider using `+kubebuilder:validation:Enum`

File: apis/config/v1alpha1/cluster_colocation_profile_types.go [hunk 2]
```diff
@@ -26,19 +26,17 @@ import (
...
 
-	// ObjectSelector decides whether to mutate/validate Pods based on if the
+	// Selector decides whether to mutate/validate Pods based on if the
 	// Pod has matching labels.
-	// Use the object selector only if the profile is opt-in, because end
-	// users may skip the admission webhook by setting the labels.
 	// Default to the empty LabelSelector, which matches everything.
 	// +optional
-	ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty"`
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
```
Comment (jasonliu747) [inline apis/config/v1alpha1/cluster_colocation_profile_types.go:39]: I think `PodSelector` might be more easy to understand. WDYT.

File: apis/config/v1alpha1/cluster_colocation_profile_types.go [hunk 3]
```diff
@@ -26,19 +26,17 @@ import (
...
-	// NamespaceSelector decides whether to mutate/validate Pods on an object based
-	// on whether the namespace for that object matches the selector.
+	// NamespaceSelector decides whether to mutate/validate Pods based
+	// on whether the namespace for that matches the selector.
 	// Default to the empty LabelSelector, which matches everything.
 	// +optional
 	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
 
-	// ObjectSelector decides whether to mutate/validate Pods based on if the
+	// Selector decides whether to mutate/validate Pods based on if the
```
Comment (jasonliu747) [inline apis/config/v1alpha1/cluster_colocation_profile_types.go:35]: Selector decides whether to mutate/validate Pods if the Pod has matching labels.

File: pkg/webhook/pod/mutating/cluster_colocation_profile.go [hunk 4]
```diff
@@ -138,6 +143,30 @@ func (h *PodMutatingHandler) doMutateByColocationProfile(pod *corev1.Pod, profil
...
+	}
+
+	if profile.Spec.PriorityClassName != "" {
+		priorityClass := &schedulingv1.PriorityClass{}
+		err := h.Client.Get(ctx, types.NamespacedName{Name: profile.Spec.PriorityClassName}, priorityClass)
+		if err != nil {
+			return err
+		}
+		pod.Spec.PriorityClassName = profile.Spec.PriorityClassName
+		pod.Spec.Priority = pointer.Int32(priorityClass.Value)
```
Comment (jasonliu747) [inline pkg/webhook/pod/mutating/cluster_colocation_profile.go:160]: there are other functions using `util.Int32`, please help me replace them with `pointer.Int32`

File: pkg/util/pod.go [hunk 5]
```diff
@@ -46,5 +46,10 @@ func GetPodRequest(pod *corev1.Pod) corev1.ResourceList {
 	if pod.Spec.Overhead != nil {
 		result = quotav1.Add(result, pod.Spec.Overhead)
 	}
-	return quotav1.Mask(result, []corev1.ResourceName{corev1.ResourceCPU, corev1.ResourceMemory})
+	return result
+}
+
+func GetPodResourcesRequest(pod *corev1.Pod, resourceNames ...corev1.ResourceName) corev1.ResourceList {
```
Comment (jasonliu747) [inline pkg/util/pod.go:52]: maybe we could combine two functions into one

File: pkg/webhook/pod/mutating/cluster_colocation_profile.go [hunk 6]
```diff
@@ -71,10 +75,11 @@ func (h *PodMutatingHandler) clusterColocationProfileMutatingPod(ctx context.Con
 	}
 
 	for _, profile := range matchedProfiles {
-		err := h.doMutateByColocationProfile(pod, profile)
+		err := h.doMutateByColocationProfile(ctx, pod, profile)
 		if err != nil {
 			return err
```
Comment (jasonliu747) [inline pkg/webhook/pod/mutating/cluster_colocation_profile.go:80]: print err

File: apis/config/v1alpha1/cluster_colocation_profile_types.go (hunk 2 above)
Comment (eahydra) [reply apis/config/v1alpha1/cluster_colocation_profile_types.go:39]: Selector is K8s API design idiom. Like StatefulSet.Spec.Selector or Deployment.Spec.Selector

File: pkg/webhook/pod/mutating/cluster_colocation_profile.go (hunk 6 above)
Comment (eahydra) [reply pkg/webhook/pod/mutating/cluster_colocation_profile.go:80]: Don't need it. Mutating handle will log the error

**Review by jasonliu747 — APPROVED** [review]
/lgtm

File: apis/config/v1alpha1/cluster_colocation_profile_types.go [hunk 7]
```diff
@@ -52,6 +50,26 @@ type ClusterColocationProfileSpec struct {
...
 	SchedulerName string `json:"schedulerName,omitempty"`
 
+	// QoSClass describes the type of koordinator QoS that the Pod is running.
+	// Can be LSE/LSR/LS/BE/SYSTEM.
+	// +kubebuilder:validation:Enum=LSE;LSR;LS;BE;SYSTEM
+	// +optional
+	QoSClass string `json:"qosClass,omitempty"`
+
+	// If specified, indicates the pod's priority.
+	// Optional can be specified koordinator-prod/koordinator-batch/koordinator-mid/koordinator-free
```
Comment (hormes) [inline apis/config/v1alpha1/cluster_colocation_profile_types.go:60]: change the order koordinator-mid/koordinator-batch

File: apis/config/v1alpha1/cluster_colocation_profile_types.go [hunk 8]
```diff
@@ -52,6 +50,26 @@ type ClusterColocationProfileSpec struct {
...
+	// default.
+	// +optional
+	PriorityClassName string `json:"priorityClassName,omitempty"`
+
+	// ColocationPriority defines the Pod priority in Koordinator.
+	// Various koordinator components use this field to find the
+	// priority of the pod.
+	// The higher the value, the higher the priority.
+	// +optional
+	ColocationPriority *int32 `json:"colocationPriority,omitempty"`
```
Comment (hormes) [inline apis/config/v1alpha1/cluster_colocation_profile_types.go:71]: Leave this field in the label?

File: apis/config/v1alpha1/cluster_colocation_profile_types.go [hunk 9]
```diff
@@ -52,6 +50,26 @@ type ClusterColocationProfileSpec struct {
...
+	// +kubebuilder:validation:Enum=LSE;LSR;LS;BE;SYSTEM
+	// +optional
+	QoSClass string `json:"qosClass,omitempty"`
+
+	// If specified, indicates the pod's priority.
+	// Optional can be specified koordinator-prod/koordinator-batch/koordinator-mid/koordinator-free
+	// If not specified, the pod priority will be default or zero if there is no
+	// default.
+	// +optional
+	PriorityClassName string `json:"priorityClassName,omitempty"`
```
Comment (hormes) [inline apis/config/v1alpha1/cluster_colocation_profile_types.go:64]: move qosClass, priorityClassName to the front of labels ?

File: apis/config/v1alpha1/cluster_colocation_profile_types.go (hunk 7 above)
Comment (jasonliu747) [reply apis/config/v1alpha1/cluster_colocation_profile_types.go:60]: if possible use `+kubebuilder:validation:Enum`

**Pull request merged**